### PR TITLE
[RLlib] Created action_dist_v2 for RLModule examples, RLModule PR 2/N

### PR DESCRIPTION
**Pull request by kouroshHakha**
Signed-off-by: Kourosh Hakhamaneshi <kourosh@anyscale.com>

## Why are these changes needed?
submitting [this pr](https://github.com/ray-project/ray/pull/29449) in pieces:
These action distributions have simpler interface and are more explicit. Previously users would have to pass in this magical input and modelv2 instance (only for auto-regressive distributions), but now the interface is more explicit and more familiar since it looks like pytorch distributions. The auto-regressive distribution can still be created by sub-classing this base-class.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/models/action_dist_v2.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
+from typing import Tuple
+import gym
+import abc
+
+from ray.rllib.utils.annotations import ExperimentalAPI
+from ray.rllib.utils.typing import TensorType, Union, ModelConfigDict
+
+
+@ExperimentalAPI
+class ActionDistributionV2(abc.ABC):
```
Comment (sven1977) [inline rllib/models/action_dist_v2.py:10]: We should really take this opportunity and get rid of the restriction to actions. Can we just call this `Distribution`?
This would then cover models that predict next states, rewards, actions, etc.. as well as serve inside variational autoencoders.

File: rllib/models/action_dist_v2.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    @abc.abstractmethod
+    def required_model_output_shape(
+        action_space: gym.Space, model_config: ModelConfigDict
+    ) -> Tuple[int, ...]:
+        """Returns the required shape of an input parameter tensor for a
+        particular action space and an optional dict of distribution-specific
+        options.
+
+        Args:
+            action_space: The action space this distribution will be used for,
```
Comment (sven1977) [inline rllib/models/action_dist_v2.py:112]: action_space -> space

File: rllib/models/action_dist_v2.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        Returns:
+            The entropy of the action distribution.
+        """
+
+    @staticmethod
+    @abc.abstractmethod
+    def required_model_output_shape(
+        action_space: gym.Space, model_config: ModelConfigDict
+    ) -> Tuple[int, ...]:
+        """Returns the required shape of an input parameter tensor for a
```
Comment (sven1977) [inline rllib/models/action_dist_v2.py:107]: Can we give 2 examples here of how this method will be used (for Categorical and DiagGaussian)?
* Does the shape include the batch dim (time dim, etc..)?
* What are typical cases, where the model config would be used and how would it be used?

File: rllib/models/action_dist_v2.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+    @abc.abstractmethod
+    def rsample(
+        self,
+        *,
+        sample_shape: Tuple[int, ...] = None,
+        return_logp: bool = False,
+        **kwargs
+    ) -> Union[TensorType, Tuple[TensorType, TensorType]]:
+        """Draw a re-parameterized sample from the action distribution.
```
Comment (sven1977) [inline rllib/models/action_dist_v2.py:55]: Can we quickly explain what the difference is to `sample`? Basically that rsample is backprop'able and sample is not (iiuc)?

File: rllib/models/action_dist_v2.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+            The sampled action. If return_logp is True, returns a tuple of the
+            sampled action and its logp.
+        """
+
+    @abc.abstractmethod
+    def rsample(
+        self,
+        *,
+        sample_shape: Tuple[int, ...] = None,
+        return_logp: bool = False,
```
Comment (sven1977) [inline rllib/models/action_dist_v2.py:52]: Great!

File: rllib/models/action_dist_v2.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+from ray.rllib.utils.annotations import ExperimentalAPI
+from ray.rllib.utils.typing import TensorType, Union, ModelConfigDict
+
+
+@ExperimentalAPI
+class ActionDistributionV2(abc.ABC):
+    """The policy action distribution of an agent.
+
+    Args:
+        inputs: input vector to define the distribution over.
```
Comment (sven1977) [inline rllib/models/action_dist_v2.py:14]: Is this always a vector? Or could this also be a dict of tensors (multi-distribution) or a >1D tensor (multi-variate diag gaussian)?

File: rllib/models/action_dist_v2.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+        self,
+        *,
+        sample_shape: Tuple[int, ...] = None,
+        return_logp: bool = False,
+        **kwargs
+    ) -> Union[TensorType, Tuple[TensorType, TensorType]]:
+        """Draw a sample from the action distribution.
+
+        Args:
+            sample_shape: The shape of the sample to draw.
```
Comment (sven1977) [inline rllib/models/action_dist_v2.py:38]: Can we be more specific here? Doesn't the input (currently `self.inputs`) already determine the batch/time dimensions and the rest is "fixed"? If this is not the case, can we add examples to this docstring?

**Review by sven1977 — COMMENTED** [review]
Awesome PR @kouroshHakha ! Just a few questions and nits on the docstrings and one renaming request.

**Comment by sven1977** [comment]
Exciting to see all these things being done-over! :)
@kouroshHakha

File: rllib/models/torch/torch_action_dist_v2.py [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+        sample = self.dist.sample(sample_shape)
+        if return_logp:
+            return sample, self.logp(sample)
+        return sample
+
+    @override(ActionDistributionV2)
+    def rsample(
+        self, *, sample_shape=torch.Size(), return_logp: bool = False
+    ) -> Union[TensorType, Tuple[TensorType, TensorType]]:
+        sample = self.dist.rsample(sample_shape)
```
Comment (sven1977) [inline rllib/models/torch/torch_action_dist_v2.py:57]: Nit: rename local var `sample` to `rsample` for clarity?

File: rllib/models/torch/torch_action_dist_v2.py [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+@DeveloperAPI
+class TorchDistribution(ActionDistributionV2, abc.ABC):
+    """Wrapper class for torch.distributions."""
+
+    def __init__(self, *args, **kwargs):
+        super().__init__()
+        self.dist = self._get_distribution(*args, **kwargs)
+
+    @abc.abstractmethod
+    def _get_distribution(self, *args, **kwargs) -> torch.distributions.Distribution:
```
Comment (sven1977) [inline rllib/models/torch/torch_action_dist_v2.py:29]: Hmm, the more I think about this, we should rename this into `_get_torch_distribution` for clarity (vs `_get_tf_distribution`).
Or even `_get_underlying_torch_distribution`. B/c `self` is already a `TorchDistribution`, albeit an RLlib one :)

File: rllib/models/torch/torch_action_dist_v2.py [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+class TorchCategorical(TorchDistribution):
+    """Wrapper class for PyTorch Categorical distribution."""
+
+    def __init__(
+        self,
+        probs: torch.Tensor = None,
+        logits: torch.Tensor = None,
+        temperature: float = 1.0,
+    ) -> None:
+        super().__init__(probs=probs, logits=logits, temperature=temperature)
```
Comment (sven1977) [inline rllib/models/torch/torch_action_dist_v2.py:73]: docstring?

File: rllib/models/torch/torch_action_dist_v2.py [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+class TorchDiagGaussian(TorchDistribution):
+    """Wrapper class for PyTorch Normal distribution."""
+
+    @override(ActionDistributionV2)
+    def __init__(
+        self,
+        loc: torch.Tensor,
+        scale: Optional[torch.Tensor] = None,
+    ):
+        super().__init__(loc=loc, scale=scale)
```
Comment (sven1977) [inline rllib/models/torch/torch_action_dist_v2.py:105]: docstring?

File: rllib/models/torch/torch_action_dist_v2.py [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+@DeveloperAPI
+class TorchDeterministic(ActionDistributionV2):
+    """Action distribution that returns the input values directly.
+
+    This is similar to DiagGaussian with standard deviation zero (thus only
+    requiring the "mean" values as NN output).
+    """
+
+    def __init__(self, loc: torch.Tensor) -> None:
+        super().__init__()
```
Comment (sven1977) [inline rllib/models/torch/torch_action_dist_v2.py:142]: docstring?

File: rllib/models/torch/torch_action_dist_v2.py [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+    @override(ActionDistributionV2)
+    def required_model_output_shape(
+        action_space: gym.Space, model_config: ModelConfigDict
+    ) -> Tuple[int, ...]:
+        return tuple(np.prod(action_space.shape, dtype=np.int32) * 2)
+
+
+@DeveloperAPI
+class TorchDeterministic(ActionDistributionV2):
+    """Action distribution that returns the input values directly.
```
Comment (sven1977) [inline rllib/models/torch/torch_action_dist_v2.py:135]: "Action distribution" -> "Distribution"

File: rllib/models/action_dist_v2.py (hunk 1 above)
Comment (kouroshHakha) [reply rllib/models/action_dist_v2.py:10]: Yeah, I agree. fixed.

File: rllib/models/action_dist_v2.py (hunk 2 above)
Comment (kouroshHakha) [reply rllib/models/action_dist_v2.py:112]: fixed.

File: rllib/models/action_dist_v2.py (hunk 3 above)
Comment (kouroshHakha) [reply rllib/models/action_dist_v2.py:107]: I am not sure if this will actually get used tbh. I have it here just as a reminder to the next developer that this was part of the old distribution that we may or may not keep depending on how the catalog gets written.  added a comment to reflect this.

File: rllib/models/action_dist_v2.py (hunk 4 above)
Comment (kouroshHakha) [reply rllib/models/action_dist_v2.py:55]: yep added.

File: rllib/models/action_dist_v2.py (hunk 6 above)
Comment (kouroshHakha) [reply rllib/models/action_dist_v2.py:14]: Oh this is a residue of the old stuff. Doesn't mean anything anymore. Removed. :)

File: rllib/models/torch/torch_action_dist_v2.py (hunk 9 above)
Comment (kouroshHakha) [reply rllib/models/torch/torch_action_dist_v2.py:29]: I agree. fixed.

File: rllib/models/action_dist_v2.py (hunk 7 above)
Comment (kouroshHakha) [reply rllib/models/action_dist_v2.py:38]: This should be addressed in the docstrings of each individual distribution.

File: rllib/models/torch/torch_action_dist_v2.py (hunk 10 above)
Comment (kouroshHakha) [reply rllib/models/torch/torch_action_dist_v2.py:73]: added

File: rllib/models/torch/torch_action_dist_v2.py (hunk 11 above)
Comment (kouroshHakha) [reply rllib/models/torch/torch_action_dist_v2.py:105]: added

File: rllib/models/torch/torch_action_dist_v2.py (hunk 12 above)
Comment (kouroshHakha) [reply rllib/models/torch/torch_action_dist_v2.py:142]: added

**Comment by kouroshHakha** [comment]
@sven1977 please re-review.

**Review by sven1977 — APPROVED** [review]
LGTM!

**Comment by sven1977** [comment]
`test_distributions_v2` failing
cc: @kouroshHakha .

**Review by gjoliver — APPROVED** [review]
I like this new base class!
just a little worried that we are not adding TF version of the implementation yet.
should be pretty easy actually?
but this is good regardless.
